### PR TITLE
move root mount setup to mountproviders

### PR DESCRIPTION
**Pull request by icewind1991**
Removes special casing for the root storage from the filesystem setup logic

Requires:

 - [x] https://github.com/nextcloud/server/pull/30978

Part of https://github.com/nextcloud/server/pull/31265

### CODE REVIEW

File: lib/private/Files/Mount/RootMountProvider.php [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+		$name = $config['class'];
+		if (strpos($name, 'OCA\\') === 0 && substr_count($name, '\\') >= 2) {
+			$segments = explode('\\', $name);
+			OC_App::loadApp(strtolower($segments[1]));
+		}
+	}
+
+	private function getLocalRootMount(IStorageFactory $loader): MountPoint {
+		$configDataDirectory = $this->config->getSystemValue("datadirectory", OC::$SERVERROOT . "/data");
+		return new MountPoint(LocalRootStorage::class, '/', ['datadir' => $configDataDirectory], $loader, null, null, self::class);
```
Comment (CarlSchwan) [inline lib/private/Files/Mount/RootMountProvider.php:74]: psalm is not happy

File: lib/private/legacy/OC_Util.php [hunk 2]
```diff
@@ -91,93 +90,6 @@ protected static function getAppManager() {
...
-	}
-
-	/**
-	 * mounting an object storage as the root fs will in essence remove the
-	 * necessity of a data folder being present.
-	 *
-	 * @param array $config containing 'class' and optional 'arguments'
-	 * @suppress PhanDeprecatedFunction
-	 */
-	private static function initObjectStoreMultibucketRootFS($config) {
```
Comment (PVince81) [inline lib/private/legacy/OC_Util.php:147]: is this also covered by unit / integration tests or needs manual retesting ?

File: lib/private/legacy/OC_Util.php [hunk 3]
```diff
@@ -91,93 +90,6 @@ protected static function getAppManager() {
 		return \OC::$server->getAppManager();
 	}
 
-	private static function initLocalStorageRootFS() {
```
Comment (PVince81) [inline lib/private/legacy/OC_Util.php:94]: glad to see those move to a less static place :smile:

File: lib/private/legacy/OC_Util.php (hunk 2 above)
Comment (icewind1991) [reply lib/private/legacy/OC_Util.php:147]: Added some tests

**Review by CarlSchwan — APPROVED** [review]
Looks good :+1: and tests are happy

File: .github/workflows/s3-primary.yml [hunk 4]
```diff
@@ -0,0 +1,70 @@
+name: S3 primary storage
```
Comment (PVince81) [inline .github/workflows/s3-primary.yml:1]: hurrah for more tests!

File: apps/dav/lib/Connector/Sabre/File.php [hunk 5]
```diff
@@ -173,7 +173,23 @@ public function put($data) {
...
+					// during very large uploads, the shared lock we got at the start might have been expired
+					// meaning that the above lock can fail not just only because somebody else got a shared lock
+					// or because there is no existing shared lock to make exclusive
+					//
+					// Thus we try to get a new exclusive lock, if the original lock failed because of a different shared
+					// lock this will still fail, if our original shared lock expired the new lock will be successful and
+					// the entire operation will be safe
+
+					try {
+						$this->acquireLock(ILockingProvider::LOCK_EXCLUSIVE);
```
Comment (PVince81) [inline apps/dav/lib/Connector/Sabre/File.php:188]: is this directly related to this PR or is this a fix that we could use separately ?

File: .github/workflows/s3-primary.yml (hunk 4 above)
Comment (PVince81) [reply .github/workflows/s3-primary.yml:1]: is this backportable separately?

**Review by PVince81 — APPROVED** [review]
Code looks fine :+1: 

It does seem to also contain changes that perhaps could be extracted to separate PRs to make them backportable if they solve bigger issues ?

**Comment by icewind1991** [comment]
I'll look into backporting the primary storage ci+fixes

**Comment by PVince81** [comment]
> I'll look into backporting the primary storage ci+fixes

thanks!

**Pull request merged**